unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java

```diff
@@ -131,24 +131,31 @@ public boolean contains(final UTF8String substring) {
     }
 
     for (int i = 0; i <= bytes.length - b.length; i++) {
-      // TODO: Avoid copying.
-      if (bytes[i] == b[0] && Arrays.equals(Arrays.copyOfRange(bytes, i, i + b.length), b)) {
+      if (bytes[i] == b[0] && startsWith(substring, i)) {
         return true;
       }
     }
     return false;
   }
 
+  private boolean startsWith(final UTF8String prefix, int offset) {
+    byte[] b = prefix.getBytes();
+    if (b.length + offset > bytes.length || offset < 0) {
+      return false;
+    }
+    int i = 0;
+    while (i < b.length && b[i] == bytes[i + offset]) {
```

Comment:
As @rxin suggested, we could use unsafe API to compare 8 bytes at once. But I think in most cases the suffix will be short and this while loop will terminate very fast, it may not worth that complication.

The while loop is good to me.